src/pubsub/ua_pubsub_writer.c

```diff
@@ -391,14 +407,65 @@ void
 UA_PublishedDataSet_clear(UA_Server *server, UA_PublishedDataSet *publishedDataSet){
     UA_PublishedDataSetConfig_clear(&publishedDataSet->config);
     //delete PDS
-    UA_DataSetMetaDataType_clear(&publishedDataSet->dataSetMetaData);
     UA_DataSetField *field, *tmpField;
     LIST_FOREACH_SAFE(field, &publishedDataSet->fields, listEntry, tmpField) {
         UA_Server_removeDataSetField(server, field->identifier);
     }
+    UA_DataSetMetaDataType_clear(&publishedDataSet->dataSetMetaData);
     UA_NodeId_clear(&publishedDataSet->identifier);
 }
 
+static UA_StatusCode
+generateFieldMetaData(UA_Server *server, UA_DataSetField *field, UA_FieldMetaData *fieldMetaData){
+    if(field->config.dataSetFieldType == UA_PUBSUB_DATASETFIELD_VARIABLE){
```

Comment:
The if clause covers the entire method .
Please early-return.

A linear code path is preferable.
Reduced complexity...

Reply:
There are different types of DataSet Fields. Therefor the metadata generation must be handeled different. I canged the big if clause to an switch/case witch handels all cases and returns UA_STATAUSCODE_BADNOTSUPPORTED if the field type is 'event'.